src/main/kotlin/org/rust/ide/refactoring/introduceConstant/RsIntroduceConstantHandler.kt

```diff
@@ -0,0 +1,99 @@
+/*
+ * Use of this source code is governed by the MIT license that can be
+ * found in the LICENSE file.
+ */
+
+package org.rust.ide.refactoring.introduceConstant
+
+import com.intellij.openapi.actionSystem.DataContext
+import com.intellij.openapi.editor.Editor
+import com.intellij.openapi.project.Project
+import com.intellij.psi.PsiDocumentManager
+import com.intellij.psi.PsiElement
+import com.intellij.psi.PsiFile
+import com.intellij.psi.PsiParserFacade
+import com.intellij.refactoring.RefactoringActionHandler
+import com.intellij.refactoring.RefactoringBundle
+import com.intellij.refactoring.util.CommonRefactoringUtil
+import org.rust.ide.inspections.import.AutoImportFix
+import org.rust.ide.refactoring.*
+import org.rust.lang.core.psi.*
+import org.rust.lang.core.psi.ext.RsElement
+import org.rust.lang.core.types.type
+import org.rust.openapiext.runWriteCommandAction
+
+class RsIntroduceConstantHandler : RefactoringActionHandler {
+    override fun invoke(project: Project, editor: Editor, file: PsiFile, dataContext: DataContext) {
+        if (file !is RsFile) return
+        val exprs = findCandidateExpressionsToExtract(editor, file).filter { it.isExtractable() }
+
+        when (exprs.size) {
+            0 -> {
+                val message = RefactoringBundle.message(if (editor.selectionModel.hasSelection())
+                    "selected.block.should.represent.an.expression"
+                else
+                    "refactoring.introduce.selection.error"
+                )
+                val title = RefactoringBundle.message("introduce.variable.title")
+                val helpId = "refactoring.extractVariable"
+                CommonRefactoringUtil.showErrorHint(project, editor, message, title, helpId)
+            }
+            1 -> extractExpression(editor, exprs.single())
+            else -> {
+                showExpressionChooser(editor, exprs) {
+                    extractExpression(editor, it)
+                }
+            }
+        }
+    }
+
+    override fun invoke(project: Project, elements: Array<out PsiElement>, dataContext: DataContext?) {
+        //this doesn't get called form the editor.
+    }
+}
+
+private fun RsExpr.isExtractable(): Boolean {
+    return when (this) {
+        is RsLitExpr -> true
+        is RsBinaryExpr -> this.left.isExtractable() && (this.right?.isExtractable() ?: true)
+        else -> false
+    }
+}
+
+private fun replaceWithConstant(expr: RsExpr, occurrences: List<RsExpr>, candidate: InsertionCandidate, editor: Editor) {
+    val project = expr.project
+    val factory = RsPsiFactory(project)
+    val suggestedNames = expr.suggestedNames()
+    val name = suggestedNames.default.toUpperCase()
+    val const = factory.createConstant(name, expr)
+
+    val (inserted, replaced) = project.runWriteCommandAction {
```

Comment:
```suggestion
    val (insertedConstent, replaced) = project.runWriteCommandAction {
```
(just to make the code a bit easier to understand)